src/main/scala/com/microsoft/hyperspace/index/PlanSignatureProvider.scala

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index
+
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+
+import com.microsoft.hyperspace.util.HashingUtils
+
+/**
+ * [[PlanSignatureProvider]] provides signature for a logical plan based on
+ * the type of operators in it.
+ * A plan needs to have at least one operator so its signature can be generated.
+ */
+class PlanSignatureProvider extends LogicalPlanSignatureProvider {
+
+  /**
+   * Generate the signature of logical plan.
+   *
+   * @param logicalPlan logical plan.
+   * @return signature if there is at least one operator in the plan; Otherwise None.
+   */
+  def signature(logicalPlan: LogicalPlan): Option[String] = {
+    var signature = ""
+    logicalPlan.foreachUp(p => signature += HashingUtils.md5Hex(p.nodeName))
```

Comment:
nit: we can avoid this calculation by just doing md5Hex in the last step in match-case statement. It would be faster and still correct. Right?

Reply:
cc @imback82

Reply:
This is an approach that we take in [FileBasedSignatureProvider](https://github.com/microsoft/hyperspace/blob/master/src/main/scala/com/microsoft/hyperspace/index/FileBasedSignatureProvider.scala#L59) as well and it makes the final value non-sensitive to the order of strings (or files in case of FileBasedSignatureProvider). This is important as for example for a plan with a  join node the signature should be independent of order of visiting join children. Otherwise the signature computation would not be reproducible if the order of visiting left and right children swap.

Reply:
@imback82 @apoorvedave1 Can you please check above comment and merge this PR if you are fine with it?

Reply:
> Visiting plan as Join,LeftChild,RightChild should produce the same signature as visiting the plan as Join,RightChild,LeftChild. Calculating Hash value per node name and summing them up makes this achievable but concatenating node names and calling hash() on final String breaks this.

@pirz are you sure about this? `hash(a) + hash(b)` is the same as `hash(b) + hash(a)` when `hash` returns a string?

@apoorvedave1 I am not too sure about the performance improvement here (<100 nodes) and would depend on how we append the strings. I would say it would be negligible. Do you have some numbers to back up?

Reply:
@imback82 My understanding of @apoorvedave1 's comment was: When visiting `a,b,c` instead of doing `hash(a) + hash(b) + hash(c)` do it as first concat `abc` and then call hash once as `hash(abc)` to save time. My point was that if we do it that way we hurt correctness as `hash(abc)` is different from `hash(acb)`. That is why we should call hash per node and add the hash values.

Reply:
Well, `hash(a) + hash(b) + hash(c)` is different from `hash(a) + hash(c) + hash(b)` since `hash` returns a string, so if there is a correctness issue in `hash(abc)`, the current approach also has an issue.

Reply:
@imback82 Actually you are right - that order does not matter if we treat the hash value as a number and add up numbers and hash the final number to a String. We do not follow that approach here and also not in `FileBasedSignatureProvider`.

Reply:
@imback82 no I don't have numbers for this, I just said it based on the understanding that hash(hash(a)+hash(b)+hash(c)) will take more time than hash(a+b+c). Other than that, my concern was the second approach looks slightly simpler and we don't lose anything. But either is fine actually.

Reply:
well, `a + b + c` involves string copies, so depending on the size of those strings, the above statement may not be true.

Reply:
OK, can we do this instead?
```scala
var signature = ""
logicalPlan.foreachUp(p => signature = HashingUtils.md5Hex(signature + p.nodeName))
signature match {
  case "" => None
  case _ => signature
}
```
We will always bound the string size and also consistent with file based signature.

Reply:
Since this PR is getting long, I will merge this now. @pirz Can you do a follow up PR to address @apoorvedave1's concern? (if we are going with one md5hex, let's use `StringBuffer`).

Reply:
Makes sense, I updated it. @imback82 and @apoorvedave1 Can you plz check/approve again?

Reply:
approved. Thanks!

Reply:
@imback82 Sure, Thanks. I guess we first need some numbers to see if calling Hash multiple times on the plans (which are normally <100 nodes as you said) is a perf bottleneck or not. If that turned out to be an issue, we will switch.